### PR TITLE
Improve Logging

### PR DESCRIPTION
**Pull request by NickCrews**
Before, if the backend had a problem when executing SQL,
you couldn't actually see the SQL that caused it unless
you had logging enabled.

This also removes the duplicated logging calls that were in
every linker subclass

This also makes it so that the logged SQL is consistent: In duckdb
and sqlite, the SQL was logged, and then further modified
by wrapping it in a "CREATE TABLE ...",
so the actual SQL ran was different
from what was logged.

### CODE REVIEW

File: splink/athena/athena_linker.py [hunk 1]
```diff
@@ -70,14 +69,12 @@ def validate(self):
         pass
 
     def drop_table_from_database(self, force_non_splink_table=False):
-
         self._check_drop_folder_created_by_splink(force_non_splink_table)
         self._check_drop_table_created_by_splink(force_non_splink_table)
         self.athena_linker.drop_table_from_database_if_exists(self.physical_name)
         self.athena_linker.delete_table_from_s3(self.physical_name)
 
     def _check_drop_folder_created_by_splink(self, force_non_splink_table=False):
```
Comment (NickCrews) [inline splink/athena/athena_linker.py:77]: Oops it looks like my black configuration is slightly different and deleted some lines. Want me to revert these changes?

File: splink/athena/athena_linker.py [hunk 2]
```diff
@@ -337,8 +326,10 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
         output_obj = self._table_to_splink_dataframe(templated_name, physical_name)
         return output_obj
 
-    def register_table(self, input, table_name, overwrite=False):
+    def _do_execute_sql(self, sql, templated_name, physical_name):
```
Comment (NickCrews) [inline splink/athena/athena_linker.py:329]: I also considered making a `@log_sql` decorator that every subclass would have to add to their `_do_execute_sql()`, instead of using the `_log_and_do_execute_sql()` framework in the superclass. But that is one more thing that each subclass would need to import and remember. I thought this was more centralized, even if it is a bit more indirection for the code reader to understand how `_log_and_do_execute_sql` delegates to `_do_execute_sql`

IDK what do you think?

File: splink/duckdb/duckdb_linker.py [hunk 3]
```diff
@@ -170,29 +166,23 @@ def _table_to_splink_dataframe(
...
         ({sql})
         """
-        self._con.execute(sql).fetch_df()
+        self._log_and_do_execute_sql(sql, templated_name, physical_name)
 
         return DuckDBLinkerDataFrame(templated_name, physical_name, self)
 
-    def register_table(self, input, table_name, overwrite=False):
+    def _do_execute_sql(self, final_sql, templated_name, physical_name):
+        self._con.execute(final_sql).fetch_df()
```
Comment (NickCrews) [inline splink/duckdb/duckdb_linker.py:183]: wait, we can just remove the .fetch_df() call, right?

Comment (NickCrews) [reply splink/duckdb/duckdb_linker.py]: removed in follow up commit

File: splink/athena/athena_linker.py (hunk 1 above)
Comment (ThomasHepworth) [reply splink/athena/athena_linker.py:77]: No need, it's fine.

File: splink/duckdb/duckdb_linker.py (hunk 3 above)
Comment (ThomasHepworth) [reply splink/duckdb/duckdb_linker.py]: Yeah, good spot. It doesn't actually materialise the whole df, it just reports back the total row count for the materialised df.

Either way, it's a legacy bit of code from several months ago that we can remove.

File: splink/linker.py [hunk 4]
```diff
@@ -469,6 +463,27 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
             f"_execute_sql_against_backend not implemented for {type(self)}"
         )
 
+    def _do_execute_sql(self, final_sql, templated_name, physical_name):
```
Comment (ThomasHepworth) [inline splink/linker.py:466]: I'd probably rename this to be `_execute_sql()`.

File: splink/linker.py [hunk 5]
```diff
@@ -469,6 +463,27 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
...
+    def _log_and_do_execute_sql(self, final_sql, templated_name, physical_name):
+        """Log the sql, then call _do_execute_sql(), wrapping any errors"""
+        logger.debug(execute_sql_logging_message_info(templated_name, physical_name))
+        logger.log(5, log_sql(final_sql))
+        try:
+            return self._do_execute_sql(final_sql, templated_name, physical_name)
+        except Error as e:
+            raise SplinkException(
+                f"""
+                Error executing sql for {templated_name} ({physical_name}):
```
Comment (ThomasHepworth) [inline splink/linker.py:482]: Possibly something such as the following is slightly clearer?

```suggestion
                Error executing the following sql for table `{templated_name}` ({physical_name}):
```

File: splink/athena/athena_linker.py (hunk 2 above)
Comment (ThomasHepworth) [reply splink/athena/athena_linker.py:329]: From conversations with Robin, we'd rather stay away from decorators as they can make the code more difficult to follow.

The method you've implemented is more in line with the code used in the wider code base.

File: splink/linker.py [hunk 6]
```diff
@@ -469,6 +463,27 @@ def _execute_sql_against_backend(self, sql, templated_name, physical_name):
...
+        """
+        raise NotImplementedError(f"_do_execute_sql not implemented for {type(self)}")
+
+    def _log_and_do_execute_sql(self, final_sql, templated_name, physical_name):
+        """Log the sql, then call _do_execute_sql(), wrapping any errors"""
+        logger.debug(execute_sql_logging_message_info(templated_name, physical_name))
+        logger.log(5, log_sql(final_sql))
+        try:
+            return self._do_execute_sql(final_sql, templated_name, physical_name)
+        except Error as e:
```
Comment (ThomasHepworth) [inline splink/linker.py:479]: I can see this is being used elsewhere in the code, but I don't believe we want to use the `Error` class from copy (here and elsewhere too...).

See the [copy docs](https://docs.python.org/3/library/copy.html).

**Comment by ThomasHepworth** [comment]
This looks great, thanks so much again for everything you've submitted recently.

I think the exception handling may be slightly awry and I can't currently get it to trigger unless I tweak it.

I've been checking this with:
```
from splink.duckdb.duckdb_linker import DuckDBLinker
import pandas as pd

from tests.basic_settings import get_settings_dict

df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")

import logging
Logger = logging.getLogger('splink').setLevel(20)

linker = DuckDBLinker(
    df,
    get_settings_dict(),
)

sql = "select count(help) from _a"
linker._enqueue_sql(sql, "testing")
linker._execute_sql_against_backend(sql, "testing", "testing")
```

I'm not very well at the moment, so apologies if I'm not the quickest in getting back to your responses.

File: splink/linker.py (hunk 6 above)
Comment (NickCrews) [reply splink/linker.py:479]: Oh oops, definitely not. Autocomplete in my IDE suggested it and I just blindly accepted it ;)

File: splink/linker.py (hunk 5 above)
Comment (NickCrews) [reply splink/linker.py:482]: Done

File: splink/linker.py (hunk 4 above)
Comment (NickCrews) [reply splink/linker.py:466]: I called it this because I don't want users to directly call this method. They should only be calling `_execute_sql_against_backend`, and then the internals of the class go and delegate to this method.

What about:

1. rename _execute_sql_against_backend -> _execute_sql(): This is what users of the class should actually be using, so make it short and obvious.
2. Keep _do_execute_sql and _log_and_do_execute_sql named as they are.

I would like to somehow hide _do_execute_sql and _log_and_do_execute_sql better so they look "More" internal than just a single leading _, but I'm not sure how to do that. I think some magix with metaclasses would work, but I think we do NOT want to go down that route. Maybe a double leading underscore? But I don't really understand how that works, some more magic name mangling happens.

**Comment by NickCrews** [comment]
I'll push up some fixes in a bit, I'm getting some errors in tests now that I removed the copy.Error catch statement...

**Comment by NickCrews** [comment]
Waiting for #1115 to land before I go back to the meat of this logging stuff

**Comment by NickCrews** [comment]
@ThomasHepworth I just pushed a version based on the new master, with a few small tweaks.

**Comment by ThomasHepworth** [comment]
This is great! Thanks so much and apologies for the delays. I've been a mixture of ill and overloaded with other priorities.

I'm going to make some minor tweaks to the [naming conventions](https://github.com/moj-analytical-services/splink/pull/690) and fix the tests, but other than that this looks great.
